Test/Unit/Model/Api/DiscountCodeValidationTest.php

```diff
@@ -740,197 +550,1661 @@ public function validateWithIncorrectWebsiteId()
+    public function applyingCouponCode()
+    {
+        $this->couponMock->expects(self::once())->method('getId')->willReturn(self::COUPON_ID);
+        $this->couponMock->expects(self::once())->method('getRuleId')->willReturn(self::RULE_ID);
+
+        $discountDesctiption = 'Test Discount';
+        $this->ruleMock->method('getDescription')->willReturn($discountDesctiption);
+        $this->ruleMock->expects(self::once())->method('getSimpleAction')->willReturn('by_percent');
+        $this->ruleMock->expects(self::once())->method('getWebsiteIds')->willReturn([self::WEBSITE_ID]);
+        $this->ruleMock->expects(self::once())->method('getRuleId')->willReturn(self::RULE_ID);
+        $this->ruleMock->expects(self::once())->method('getToDate')
+            ->willReturn(date('Y-m-d', strtotime('tomorrow')));
+        $this->ruleMock->expects(self::once())->method('getFromDate')
+            ->willReturn(date('Y-m-d', strtotime('yesterday')));
+
+        $shippingDiscountAmount = 5;
+        $this->shippingAddressMock->method('getDiscountAmount')->willReturn($shippingDiscountAmount);
+
+        $parentQuoteMock = $this->getQuoteMock(
+            self::COUPON_CODE,
+            null,
+            null ,
+            false,
+            self::PARENT_QUOTE_ID,
+            self::PARENT_QUOTE_ID
+        );
+        $immutableQuoteMock = $this->getQuoteMock(self::COUPON_CODE);
+
+        $expected = [
+            'status'          => 'success',
+            'discount_code'   => self::COUPON_CODE,
+            'discount_amount' => $shippingDiscountAmount * 100,
+            'description'     => 'Discount ' . $discountDesctiption,
+            'discount_type'   => 'percentage',
+        ];
+
+        $result = $this->invokeNonAccessibleMethod(
+            'applyingCouponCode', [
+                self::COUPON_CODE,
+                $this->couponMock,
+                $immutableQuoteMock,
+                $parentQuoteMock
+            ]
+        );
+
+        self::assertEquals($expected, $result);
+    }
+
+    /**
+     * @test
+     */
+    public function applyingCouponCode_noSuchEntity()
+    {
+        $this->couponMock->expects(self::once())->method('getRuleId')
+            ->willThrowException(new NoSuchEntityException());
+
+        $this->expectErrorResponse(
+            BoltErrorResponse::ERR_CODE_INVALID,
+            sprintf('The coupon code %s is not found', self::COUPON_CODE),
+            404
+        );
+
+        $this->assertFalse(
+            $this->invokeNonAccessibleMethod(
+                'applyingCouponCode', [
+                    self::COUPON_CODE,
+                    $this->couponMock,
+                    $this->immutableQuoteMock,
+                    $this->parentQuoteMock
+                ]
+            )
+        );
+    }
+
+    /**
+     * @test
+     */
+    public function applyingCouponCode_invalidToDate()
+    {
+        $this->couponMock->expects(self::once())->method('getId')->willReturn(self::COUPON_ID);
```

Comment:
For these coupon code tests, there seems to be a lot of repeated code. Can we create a generic function to set the mocks and then overwrite the necessary components for each test?

Reply:
@ethanwayda22 done as in the link https://github.com/BoltApp/bolt-magento2/blob/discount-code-validation-unit-tests/Test/Unit/Model/Api/DiscountCodeValidationTest.php#L255